internal/config/pipeline.go

```diff
@@ -0,0 +1,194 @@
+package config
+
+import (
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/gohcl"
+
+	pb "github.com/hashicorp/waypoint/pkg/server/gen"
+)
+
+// Pipeline represents a single pipeline definition
+type Pipeline struct {
+	Id   string `hcl:",label"`
+	Name string `hcl:"name,optional"`
+
+	StepRaw []*hclStage `hcl:"step,block"`
+
+	Body hcl.Body `hcl:",body"`
+
+	ctx    *hcl.EvalContext
+	config *Config
+}
+
+type hclPipeline struct {
+	Id   string `hcl:",label"`
+	Name string `hcl:"name,optional"`
+
+	// We need these raw values to determine the plugins need to be used.
+	StepRaw []*hclStage `hcl:"step,block"`
+
+	Body   hcl.Body `hcl:",body"`
+	Remain hcl.Body `hcl:",remain"`
+}
+
+// Pipelines returns the id of all the defined pipelines
+func (c *Config) Pipelines() []string {
+	var result []string
+	for _, p := range c.hclConfig.Pipelines {
+		result = append(result, p.Id)
+	}
+
+	return result
+}
+
+// Pipeline returns the configured pipeline named n. If the pipeline doesn't
+// exist, this will return (nil, nil).
+func (c *Config) Pipeline(id string, ctx *hcl.EvalContext) (*Pipeline, error) {
+	ctx = appendContext(c.ctx, ctx)
+
+	// Find the pipeline by progressively decoding
+	var rawPipeline *hclPipeline
+	for _, p := range c.hclConfig.Pipelines {
+		if p.Id == id {
+			rawPipeline = p
+			break
+		}
+	}
+	if rawPipeline == nil {
+		return nil, nil
+	}
+
+	// Full decode
+	var pipeline Pipeline
+	if diag := gohcl.DecodeBody(rawPipeline.Body, finalizeContext(ctx), &pipeline); diag.HasErrors() {
+		return nil, diag
+	}
+	pipeline.Id = rawPipeline.Id
+	pipeline.Name = rawPipeline.Name
+	pipeline.ctx = ctx
+	pipeline.config = c
+	if pipeline.config != nil {
+		pipeline.config.ctx = ctx
+	}
+
+	return &pipeline, nil
+}
+
+// Ref returns the ref for this pipeline.
+func (c *Pipeline) Ref() *pb.Ref_Pipeline {
+	return &pb.Ref_Pipeline{
+		Ref: &pb.Ref_Pipeline_Id{
+			Id: &pb.Ref_PipelineId{
+				Id: c.Id,
+			},
+		},
+	}
+}
+
+// Step loads the associated section of the configuration
+func (c *Pipeline) Step(ctx *hcl.EvalContext) ([]*Step, error) {
```

Comment:
I think we can parse all the steps when we parse the pipeline, no need to lazily do it.